### PR TITLE
fix!: bug 4157, use getAccounts on HD Keyring when calling addNewAccount

### LINKED ISSUE #4157: KeyringController addAccounts retrieves all addresses instead of only hd addresses
When invoking `addAccount` in the keyring controller, it calls `this.getAccounts` to fetch all addresses across every keyring, rather than limiting to just the addresses from the HD keyring. This results in a bug when accounts from various keyring types are added to a client. Consequently, the count of accounts obtained from `getAccounts` exceeds the actual number of HD accounts, leading to instances where the returned address is undefined. This issue may contribute to the selectedAddress being undefined in clients.

```
async addNewAccount(accountCount?: number): Promise<{
    keyringState: KeyringControllerMemState;
    addedAccountAddress: string;
  }> {
    const primaryKeyring = this.getKeyringsByType('HD Key Tree')[0] as
      | EthKeyring<Json>
      | undefined;
    if (!primaryKeyring) {
      throw new Error('No HD keyring found');
    }
    const oldAccounts = await this.getAccounts();

    if (accountCount && oldAccounts.length !== accountCount) {
      if (accountCount > oldAccounts.length) {
        throw new Error('Account out of sequence');
      }
      // we return the account already existing at index `accountCount`
      const primaryKeyringAccounts = await primaryKeyring.getAccounts();
      return {
        keyringState: this.#getMemState(),
        addedAccountAddress: primaryKeyringAccounts[accountCount], <-- this would be undefined.
      };
    }
```

### PR DESCRIPTION
**Pull request by montelaidev**
## Explanation

This PR addresses a bug in the keyring controller where invoking addNewAccount inadvertently retrieves addresses from all keyrings, rather than exclusively from the HD keyring. This behavior leads to inconsistencies and errors when accounts from different keyring types are present. The primary symptom of this bug is the retrieval of an incorrect number of accounts, with some operations potentially resulting in an undefined address. This issue is notably impactful as it can cause the selectedAddress to be undefined in clients, leading to unexpected behavior and degraded user experience.

## References

* Fixes #4157

### `@metamask/keyring-controller`

- **BREAKING**: Use getAccounts on HD keyring when calling `addNewAccount` instead of `this.getAccounts` which retrieves every account. 

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/keyring-controller/src/KeyringController.test.ts [hunk 1]
```diff
@@ -176,6 +176,32 @@ describe('KeyringController', () => {
...
+      await withController(
+        { keyringBuilders: [keyringBuilderFactory(MockKeyring)] },
+        async ({ controller, initialState }) => {
+          await controller.addNewKeyring(MockKeyring.type);
+
+          // expect there to be two accounts, 1 from HD and 1 from MockKeyring
+          expect(await controller.getAccounts()).toHaveLength(2);
+
+          const accountCount = initialState.keyrings[0].accounts.length;
+          const { addedAccountAddress: firstAccountAdded } =
```
Comment (ccharly) [inline packages/keyring-controller/src/KeyringController.test.ts:193]: Let's add a comment:
```ts
// We add a new account for "index 2" (not existing yet)
```

File: packages/keyring-controller/src/KeyringController.test.ts [hunk 2]
```diff
@@ -176,6 +176,32 @@ describe('KeyringController', () => {
...
+            await controller.addNewAccount(accountCount);
+          const { addedAccountAddress: secondAccountAdded } =
```
Comment (ccharly) [inline packages/keyring-controller/src/KeyringController.test.ts:195]: And another one here:
```ts
// Adding an account for an existing index will return the existing account's address
```

File: packages/keyring-controller/src/KeyringController.ts [hunk 3]
```diff
@@ -573,17 +573,22 @@ export class KeyringController extends BaseController<
     if (!primaryKeyring) {
       throw new Error('No HD keyring found');
     }
-    const oldAccounts = await this.getAccounts();
+    const oldAccounts = await primaryKeyring.getAccounts();
```
Comment (mikesposito) [inline packages/keyring-controller/src/KeyringController.ts:576]: If we change this here, then we probably have to change what's being used by the extension as `accountCount`, as we currently count _all_ accounts there: https://github.com/MetaMask/metamask-extension/blob/6e0483da7d5269cbb3afdba2ea579902e9d2987b/app/scripts/metamask-controller.js#L3602

Comment (mikesposito) [reply packages/keyring-controller/src/KeyringController.ts:576]: for this reason we should mark this PR as breaking

Comment (montelaidev) [reply packages/keyring-controller/src/KeyringController.ts:576]: Good point, we should mark this as breaking and also make the change in the clients as well.

**Comment by montelaidev** [comment]
@metamaskbot publish-previews

**Review by ccharly — APPROVED** [review]

**Pull request merged**